Source/Particles/ParticleCreation/ElementaryProcess.H

```diff
@@ -212,12 +212,11 @@ public:
         amrex::Gpu::ManagedDeviceVector<int> v_pid_product,
         amrex::Gpu::ManagedDeviceVector<WarpXParticleContainer::ParticleType*> v_particles_product,
         WarpXParticleContainer::ParticleType* particles_source,
-        const amrex::Vector<copyParticle>& v_copy_functor,
+        const amrex::Gpu::ManagedDeviceVector<copyParticle>& v_copy_functor,
         amrex::GpuArray<int*,1> runtime_iattribs_source)
     {
         int const * const AMREX_RESTRICT p_is_flagged = is_flagged.dataPtr();
         int const * const AMREX_RESTRICT p_i_product = i_product.dataPtr();
-        copyParticle const * const p_copy_functor = v_copy_functor.data();
```

Comment:
I am now copying the `ManagedDeviceVector` over to the device. We could probably also just give the `dataPtr()` - is that preferred @atmyers ?

Reply:
I'm surprised that this works - I guess I don't understand what cuda does when we launch a kernel. Maybe it constructs a deep copy of the Vector on the host in a staging area, the memcopies that to the device? `amrex::Vector` doesn't have any `__device__` constructors.

Reply:
this is not a `amrex::Vector` anymore but a `amrex::ManagedDeviceVector`. I imagine that the simple default copy constructor is called here which copies over sizes and pointer locations and the data behind the pointers is migrated as usual managed memory.

Reply:
Actually, If I compile with `DEBUG=TRUE` I get the following error message:
```
./Source/Particles/ParticleCreation/ElementaryProcess.H(238): error: calling a __host__ function("amrex::PODVector<    ::copyParticle,  ::amrex::ManagedArenaAllocator<    ::copyParticle> > ::operator [] const") from a __device__ function(" const") is not allowed

./Source/Particles/ParticleCreation/ElementaryProcess.H(238): error: identifier "amrex::PODVector<    ::copyParticle,  ::amrex::ManagedArenaAllocator<    ::copyParticle> > ::operator [] const" is undefined in device code

./Source/Particles/ParticleCreation/ElementaryProcess.H(247): warning: calling a __host__ function("amrex::PODVector<    ::copyParticle,  ::amrex::ManagedArenaAllocator<    ::copyParticle> > ::PODVector") from a __host__ __device__ function("") is not allowed

./Source/Particles/ParticleCreation/ElementaryProcess.H(247): warning: calling a __host__ function("amrex::PODVector<    ::copyParticle,  ::amrex::ManagedArenaAllocator<    ::copyParticle> > ::PODVector") from a __host__ __device__ function("") is not allowed

./Source/Particles/ParticleCreation/ElementaryProcess.H(247): warning: calling a __host__ function("amrex::PODVector<    ::copyParticle,  ::amrex::ManagedArenaAllocator<    ::copyParticle> > ::~PODVector") from a __host__ __device__ function("") is not allowed
```
This message does not show up if I compile with `DEBUG=FALSE`. However, I am not sure that the issue goes away...

Reply:
Yes, that's what I meant in

> Actually, let me update this. I do not like it as it's passing the whole container now in terms of safety/documented cuda behavior.


Update pushed now :)

Reply:
Looks good now :)